corehq/apps/smsforms/models.py

```diff
@@ -237,3 +240,82 @@ def move_to_next_action(self):
         while self.current_action_is_a_reminder and self.current_action_due < utcnow():
             self.current_reminder_num += 1
             self.set_current_action_due_timestamp()
+
+
+class XFormsSessionSynchronization:
+
+    @classmethod
+    def claim_phone_number_for_session(cls, phone_number, session):
+        """
+        This is a non-blocking acquire.
+
+        The session that claims a phone number must also release it when it's over.
+        Returns True if it was able to claim the number.
+        """
+        with cls._critical_section(phone_number):
+            if cls.could_maybe_claim_phone_number_for_session_id(phone_number, session.session_id):
+                cls._set_running_session_info_for_phone_number(
+                    phone_number,
+                    RunningSessionInfo(session.session_id, session.connection_id),
+                    # We aren't relying on the expiry here to free up the session: we manually release it.
+                    # Still, there is no situation where we'd want to keep this longer than that.
+                    session.expire_after * 60,
+                )
+                return True
+            else:
+                return False
+
+    @classmethod
+    def could_maybe_claim_phone_number_for_session_id(cls, phone_number, session_id):
+        """
+        Check if there's another session running for heuristic purposes
+
+        Does not actually claim the number, or guarantee that the session could.
+        A subsequent call to claim_phone_number_for_session could still return False.
+        """
+        running_session_info = cls.get_running_session_info_for_phone_number(phone_number)
+        return not running_session_info.session_id or running_session_info.session_id == session_id
+
+    @classmethod
+    def release_phone_number_from_session(cls, phone_number, session):
+        """
+        This allows another session to claim the phone number
+
+        The contact_id remains set, and incoming SMS will keep affinity with that contact
+        until another session claims the phone number.
+        """
+        with cls._critical_section(phone_number):
+            running_session_info = cls.get_running_session_info_for_phone_number(phone_number)
+            if cls.could_maybe_claim_phone_number_for_session_id(phone_number, session.session_id):
+                # Drop the session_id but keep the contact_id
+                # This will let incoming SMS keep affinity with that contact_id until a new session starts
+                running_session_info = running_session_info._replace(session_id=None)
+                cls._set_running_session_info_for_phone_number(
+                    phone_number, running_session_info,
+                    # Keep affinity for 30 days
```

Comment:
Would there be a downside to keeping affinity forever?

Reply:
Only a small downside of keeping more things in the cache. Not sure if this is the right tradeoff, but all numbers here seem arbitrary, and "forever" seems wasteful. Not sure how wasteful it would really be in practice, but that was my thinking

Reply:
Sounds reasonable. I agree all numbers seem arbitrary.